SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,274 @@
+
+	VirtualWorkerId CurrentVirtualWorkerId = SpatialConstants::INVALID_VIRTUAL_WORKER_ID + 1;
+
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		check(Settings->bEnableUnrealLoadBalancer);
+		const ASpatialWorldSettings* WorldSettings = GetWorld() ? Cast<ASpatialWorldSettings>(GetWorld()->GetWorldSettings()) : nullptr;
+
+		if (WorldSettings == nullptr)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, WorldSettings should inherit from SpatialWorldSettings to get the load balancing strategy."));
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+			DefaultLBStrategy->AddToRoot();
+			LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+			LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+			return;
+		}
+
+		// This will be uncommented after the next PR.
+		// For each Layer, add a LB Strategy for that layer.
+// 			TMap<FName, FLBLayerInfo> WorkerLBLayers = WorldSettings->WorkerLBLayers;
+// 
+// 			for (const TPair<FName, FLayerInfo>& Layer : Settings->WorkerLayers)
+// 			{
+// 				FName LayerName = Layer.Key;
+// 
+// 				// Look through the WorldSettings to find the LBStrategy type for this layer.
+// 				if (!WorkerLBLayers.Contains(LayerName))
+// 				{
+// 					UE_LOG(LogLayeredLBStrategy, Error, TEXT("Layer %s does not have a defined LBStrategy in the WorldSettings. It will not be simulated."), *(LayerName.ToString()));
+// 					continue;
+// 				}
+// 
+// 				UAbstractLBStrategy* LBStrategy = NewObject<UAbstractLBStrategy>(this, WorkerLBLayers[LayerName].LoadBalanceStrategy);
+// 				LayerNameToLBStrategy.Add(LayerName, LBStrategy);
+// 				LBStrategy->Init();
+// 				LBStrategy->AddToRoot();
+// 
+// 				for (const TSoftClassPtr<AActor>& ClassPtr : Layer.Value.ActorClasses)
+// 				{
+// 					ClassPathToLayer.Add(ClassPtr, LayerName);
+// 				}
+// 			}
+
+		// Finally, add the default layer.
+		if (WorldSettings->LoadBalanceStrategy == nullptr)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, there must be a LoadBalancing strategy set. Using a 1x1 grid."));
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+			DefaultLBStrategy->AddToRoot();
+			LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+			LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+		}
+		else
+		{
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UAbstractLBStrategy>(this, WorldSettings->LoadBalanceStrategy);
+			DefaultLBStrategy->AddToRoot();
+			LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+			LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+		}
+	}
+}
+
+void ULayeredLBStrategy::SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId)
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
```

Comment:
One thing... We are using const auto& Elem but calling a non const function (SetLocalVirtualWorkerId) on one of it's members. Might not want to use const here even though it doesn't actually make a difference.

Reply:
The Elem here is actually a pair, and that pair is const, even if we modify one of the things pointed to by the pointer in the pair.

Reply:
Indeed it is a cont pair ref. I was just interested in whether it's practice to make objects const even though we want to modify a member on that object. I know it is a very small detail, just interested in what your opinion is.

Reply:
I don't personally have a strong opinion on the const-ness of it. Mostly I like it const because if it's not (as you noticed below) there's a natural question in a reader of why is it not const.